asQ/ensemble.py

```diff
@@ -42,35 +43,82 @@ def split_ensemble(ensemble, split_size, **kwargs):
+    weakref.finalize(new_ensemble, split_ensemble_comm.Free)
+
+    return new_ensemble
+
+
+class ManualEnsemble(Ensemble):
+    def __init__(self, global_comm, spatial_comm, ensemble_comm, **kwargs):
         """
-        An Ensemble created from provided spatial communicators (ensemble.comm).
+        An Ensemble created from provided comms.
 
         :arg global_comm: global communicator the Ensemble is defined over.
-        :arg local_comm: communicator to use for the Ensemble.comm member.
-        :arg nmembers: number of Ensemble members (ensemble.ensemble_comm.size).
+        :arg spatial_comm: communicator to use for the Ensemble.comm member.
+        :arg ensemble_comm: communicator to use for the Ensemble.ensemble_comm member.
+
+        The global_comm, spatial_comm, and ensemble_comm must have the same logical meaning
+        as they do in firedrake.Ensemble. i.e. the global_comm is the union of a cartesian
+        product of multiple spatial_comms and ensemble_comms.
+            - ManualEnsemble is logically defined over all ranks in global_comm.
+            - Each rank in global_comm belongs to only one spatial_comm and one ensemble_comm.
+            - The size of the intersection of any (spatial_comm, ensemble_comm) pair is 1.
+
+        WARNING: Not meeting these requirements may produce in errors, hangs, and nonsensical results.
+
+        ManualEnsemble will not Free any of the comms. This is the responsibility of the user.
         """
-        if nmembers*local_comm.size != global_comm.size:
-            msg = "The global ensemble must have the same number of ranks as the sum of the local comms"
-            raise ValueError(msg)
+        # are we handed user comms?
+
+        for comm in (global_comm, spatial_comm, ensemble_comm):
+            if is_pyop2_comm(comm):
+                raise PyOP2CommError("Cannot construct Ensemble from PyOP2 internal comm")
+
+        # check cartesian product consistency
+
+        if spatial_comm.size*ensemble_comm.size != global_comm.size:
+            msg = "The global comm must have the same number of ranks as the product of spatial and ensemble comms"
+            raise PyOP2CommError(msg)
+
+        global_group = global_comm.Get_group()
+        spatial_group = spatial_comm.Get_group()
+        ensemble_group = ensemble_comm.Get_group()
+
+        if MPI.Group.Intersection(spatial_group, ensemble_group).size != 1:
+            raise PyOP2CommError("spatial and ensemble comms must be cartesian product in global_comm")
+
+        is_subgroup = lambda sub, group: MPI.Group.Compare(sub, MPI.Group.Intersection(sub, group)) in {MPI.IDENT, MPI.CONGRUENT}
+
+        if not is_subgroup(spatial_group, global_group):
+            raise PyOP2CommError("spatial_comm must be subgroup of global_comm")
+        if not is_subgroup(ensemble_group, global_group):
+            raise PyOP2CommError("ensemble_comm must be subgroup of global_comm")
+
```

Comment:
This logic isn't completely exhaustive, it doesn't currently check whether you have the same communicator. For example:
```python
if __name__ == "__main__":
    r = COMM_WORLD.rank
    s = COMM_WORLD.size

    ensemble_color1 = int(r < s/2)
    ensemble1 = COMM_WORLD.Split(color=ensemble_color1, key=r)

    ensemble_color2 = r >= s/2
    ensemble2 = COMM_WORLD.Split(color=ensemble_color2, key=r)

    spatial_color = r % (s/2)
    spatial = COMM_WORLD.Split(color=spatial_color, key=r)

    correct = ManualEnsemble(COMM_WORLD, spatial, ensemble1)
    if r < s/2:
        broken = ManualEnsemble(COMM_WORLD, spatial, ensemble1)
    else:
        broken = ManualEnsemble(COMM_WORLD, spatial, ensemble2)

    print("ALL PASSED")
```
Will run just fine, but the `broken` ensemble uses two different communicators.

This is broken in a very subtle way as the mismatched comm will be destroyed when the ensemble is destroyed.

Reply:
I don't see what is broken about this.

Assuming 8 ranks, ranks 0-3 use the ensemble comm from the first `Split`, and ranks 4-7 use the ensemble comm from the second `Split` call, but this doesn't matter. All ranks in each comm use the same one, and they don't "know", and don't need to, what the other half is doing, so long as every rank has an ensemble comm that connects the same part of all spatial comms.

As written so far, `ManualEnsemble` doesn't destroy any of the comms its given (the docstring explicitly says that the caller is responsible for this).